False positive for `inconsistent-return-statements` with method annotated with `NoReturn` return type

### Bug description

`repro.py`
```python
from typing import NoReturn

class Foo:
    def _never_returns() -> NoReturn:
        raise Exception

    def bar(self) -> int:
        try:
            return 1
        except Exception:
            self._never_returns()
```

### Configuration

```ini
Default Config
```


### Command used

```shell
pylint --disable=all --enable=inconsistent-return-statements repro.py
```


### Pylint output

```shell
************* Module repro
repro.py:7:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 0.00/10, +8.89)
```


### Expected behavior

```console
------------------------------------
Your code has been rated at 10.00/10
```

### Pylint version

```shell
My pylint configuration is pinned to 


pylint 2.15.8
astroid 2.13.5
Python 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]


But I was able to repro this on the latest released version of pylint (2.17.4 at time of writing).
```


### OS / Environment

```console
$ lsb_release -a
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.2 LTS
Release:	22.04
Codename:	jammy
$ uname -r
5.15.0-72-generic
```



### Additional dependencies

_No response_


---

I did an initial investigation, and _might_ have an idea of what's going on:


* During the `inconsistent-return-statements` lint, there's an attempt made infer whether the return type of a function call is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L2001-L2016

* But `_is_function_def_never_returning` only inspects the return type if the node passed in is a `nodes.FunctionDef`. But calling `node.func.inferred()[0]` on the method call (e.g. `self._never_returns()`) returns the corresponding method definition node as type `astroid.BoundMethod`. Since that isn't a `nodes.FunctionDef`, `_is_function_def_never_returning` never even looks to see if the return type is `NoReturn`.

  https://github.com/pylint-dev/pylint/blob/a67ffbdd261dffa01045162506614bd8c1e21be5/pylint/checkers/refactoring/refactoring_checker.py#L1956-L1957

--- a/doc/whatsnew/fragments/8747.false_positive
+++ b/doc/whatsnew/fragments/8747.false_positive
@@ -1,0 +1,3 @@
+Fix a false positive where pylint was ignoring method calls annotated as ``NoReturn`` during the ``inconsistent-return-statements`` check.
+
+Closes #8747

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -1998,16 +1998,18 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             next_sibling = next_sibling.next_sibling()
         return False
 
-    def _is_function_def_never_returning(self, node: nodes.FunctionDef) -> bool:
+    def _is_function_def_never_returning(
+        self, node: nodes.FunctionDef | astroid.BoundMethod
+    ) -> bool:
         """Return True if the function never returns, False otherwise.
 
         Args:
-            node (nodes.FunctionDef): function definition node to be analyzed.
+            node (nodes.FunctionDef or astroid.BoundMethod): function definition node to be analyzed.
 
         Returns:
             bool: True if the function never returns, False otherwise.
         """
-        if isinstance(node, nodes.FunctionDef) and node.returns:
+        if isinstance(node, (nodes.FunctionDef, astroid.BoundMethod)) and node.returns:
             return (
                 isinstance(node.returns, nodes.Attribute)
                 and node.returns.attrname == "NoReturn"

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
@@ -84,9 +84,9 @@ class ClassUnderTest:
         except ValueError:
             self._does_return_method()
 
-    def bug_pylint_8747_incorrect_annotation_trumps_inference(self, s: str) -> int:
-        """Every return is consistent since pylint does not currently use inference
-        to detect that the NoReturn annotation is incorrect and the function actually returns
+    def bug_pylint_8747_incorrect_annotation(self, s: str) -> int:
+        """Every return is consistent since pylint does not attempt to detect that the
+        NoReturn annotation is incorrect and the function actually returns
         """
         try:
             n = int(s)

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
@@ -58,6 +58,9 @@ class ClassUnderTest:
     def _no_return_method(self) -> typing.NoReturn:
         sys.exit(1)
 
+    def _falsely_no_return_method(self) -> typing.NoReturn:
+        return 1
+
     def _does_return_method(self) -> int:
         return 1
 
@@ -80,3 +83,16 @@ class ClassUnderTest:
             return n
         except ValueError:
             self._does_return_method()
+
+    def bug_pylint_8747_underhanded(self, s: str) -> int:
+        """Every return is not consistent because self._falsely_no_return_method() returns a value.
+
+        Known false negative because we take the typing into account and not inference
+        contrary to what we usually do."""
+        try:
+            n = int(s)
+            if n < 1:
+                raise ValueError
+            return n
+        except ValueError:
+            self._falsely_no_return_method()

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
@@ -84,11 +84,10 @@ class ClassUnderTest:
         except ValueError:
             self._does_return_method()
 
-    def bug_pylint_8747_underhanded(self, s: str) -> int:
-        """Every return is not consistent because self._falsely_no_return_method() returns a value.
-
-        Known false negative because we take the typing into account and not inference
-        contrary to what we usually do."""
+    def bug_pylint_8747_incorrect_annotation_trumps_inference(self, s: str) -> int:
+        """Every return is consistent since pylint does not currently use inference
+        to detect that the NoReturn annotation is incorrect and the function actually returns
+        """
         try:
             n = int(s)
             if n < 1:

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.py
@@ -53,3 +53,30 @@ def bug_pylint_4122_bis(s):
         return n
     except ValueError:
         parser_error_name('parser error')
+
+class ClassUnderTest:
+    def _no_return_method(self) -> typing.NoReturn:
+        sys.exit(1)
+
+    def _does_return_method(self) -> int:
+        return 1
+
+    def bug_pylint_8747(self, s: str) -> int:
+        """Every return is consistent because self._no_return_method hints NoReturn"""
+        try:
+            n = int(s)
+            if n < 1:
+                raise ValueError
+            return n
+        except ValueError:
+            self._no_return_method()
+
+    def bug_pylint_8747_wrong(self, s: str) -> int:  # [inconsistent-return-statements]
+        """Every return is not consistent because self._does_return_method() returns a value"""
+        try:
+            n = int(s)
+            if n < 1:
+                raise ValueError
+            return n
+        except ValueError:
+            self._does_return_method()

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.txt
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.txt
@@ -1,2 +1,2 @@
 inconsistent-return-statements:32:0:32:25:bug_pylint_4122_wrong:Either all return statements in a function should return an expression, or none of them should.:UNDEFINED
-inconsistent-return-statements:74:4:74:29:ClassUnderTest.bug_pylint_8747_wrong:Either all return statements in a function should return an expression, or none of them should.:UNDEFINED
+inconsistent-return-statements:77:4:77:29:ClassUnderTest.bug_pylint_8747_wrong:Either all return statements in a function should return an expression, or none of them should.:UNDEFINED

--- a/tests/functional/i/inconsistent/inconsistent_returns_noreturn.txt
+++ b/tests/functional/i/inconsistent/inconsistent_returns_noreturn.txt
@@ -1,1 +1,2 @@
 inconsistent-return-statements:32:0:32:25:bug_pylint_4122_wrong:Either all return statements in a function should return an expression, or none of them should.:UNDEFINED
+inconsistent-return-statements:74:4:74:29:ClassUnderTest.bug_pylint_8747_wrong:Either all return statements in a function should return an expression, or none of them should.:UNDEFINED

